### PR TITLE
SMAC tuner bug fixes

### LINKED ISSUE #1641: SMAC cannot have lower bound equal to upper bound
Found by #1570.

SMAC does not allow lower bound equal to upper bound. The problem exists for randint, uniform, quniform and loguniform.

Seems that intercepting these kinds of search space with nni is a valid fix.

### PR DESCRIPTION
**Pull request by ultmaster**
1. Fix #1641.
2. Fix the problem that `update_search_space` can modify the object of search space.
3. Set default value of optimize mode to be `maximize`.

Tests in #1570 passed.

### CODE REVIEW

File: src/sdk/pynni/nni/smac_tuner/smac_tuner.py [hunk 1]
```diff
@@ -129,23 +131,39 @@ def _main_cli(self):
...
+        # TODO: this is ugly, we put all the initialization work in this method, because initialization relies
+        #         on search space, also because update_search_space is called at the beginning.
+
         if not self.update_ss_done:
-            self.categorical_dict = generate_scenario(search_space)
+            updated_search_space = dict()
+            self.fixed_value_set = dict()
+            for key, val in search_space.items():
+                try:
+                    if "_type" in val and ((val["_type"] == "randint" and val["_value"][0] + 1 == val["_value"][1])
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/smac_tuner/smac_tuner.py]: too ugly

Comment (ultmaster) [reply src/sdk/pynni/nni/smac_tuner/smac_tuner.py]: What's your suggestion then?

Comment (QuanluZhang) [reply src/sdk/pynni/nni/smac_tuner/smac_tuner.py]: please write this logic in `generate_pcs`, convert them to `categorical `.

File: src/sdk/pynni/nni/smac_tuner/convert_ss_to_scenario.py [hunk 2]
```diff
@@ -42,72 +41,72 @@ def get_json_content(file_path):
...
-                                json.dumps(search_space[key]['_value'][0])))
+                            # use np.round here to ensure that the rounded default value is in the range,
+                            # which will be rounded in configure_space package
+                            low, high = list(np.round(np.log(search_space[key]['_value']), 10))
+                            if low == high:
+                                dump_categorical(pcs_fd, key, [search_space[key]['_value'][0]])
+                            else:
+                                pcs_fd.write('%s real [%s, %s] [%s]\n' % (key, low, high, low))
                         elif search_space[key]['_type'] == 'quniform':
                             low, high, q = search_space[key]['_value'][0:3]
```
Comment (suiguoxin) [inline src/sdk/pynni/nni/smac_tuner/convert_ss_to_scenario.py]: is it possible that low == high here ?

**Pull request merged**